### PR TITLE
JavaScript: Port the Ruby time complexity lesson to JavaScript course

### PR DESCRIPTION
**Pull request by 01zulfi**
## Because

Ruby course has a new DSA lesson on time complexity: https://github.com/TheOdinProject/curriculum/pull/24888

## This PR

Translates the Ruby lesson for JavaScript course

## Issue

n/a

## Pull Request Requirements

-   [X] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [X] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [X] The `Because` section summarizes the reason for this PR
-   [X] The `This PR` section has a bullet point list describing the changes in this PR
-   [X] If this PR addresses an open issue, it is linked in the `Issue` section
-   [X] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [X] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

**Comment by 01zulfi** [comment]
@CouchofTomato I've made the bare minimum changes: converted the Ruby code examples to JavaScript and replaced a couple of Assignment items. If those look good, you can merge this PR whenever.

File: javascript/computer_science/time_complexity.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+- O(n&#179;) - Cubic Complexity
+- O(2&#8319;) - Exponential Complexity
+- O(N!) - Factorial Complexity
+
+#### O(1) - Constant Complexity
+
+To understand Constant Complexity let's use a simple array.
+
+~~~js
+arr = [1, 2, 3, 4, 5]
```
Comment (CouchofTomato) [inline javascript/computer_science/time_complexity.md:123]: Are we not declaring with `let` or anything for the JS part?

File: javascript/computer_science/time_complexity.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+arr = [6, 7, 8, -, -]
+~~~
+
+We can keep doing this until we have an array of just one item. If it matches the number we're looking for we've found it, if not then it isn't in the array.
+
+The below table summarises the size of an array doubling and how many steps in Big O terms we would need to arrive at one element to see if it matches what we're looking for
+
+| Size | Steps |
+| ----------- | ----------- |
+| 1 | 0 |
```
Comment (CouchofTomato) [inline javascript/computer_science/time_complexity.md:168]: I think in the Ruby part I made a mistake here that's carried over. I think size 1 would be 1 step. size 2 would be 2 then 3 would be 4. Not sure what I was thinking but I don't think an alogrithm can have zero steps unless it's empty. Sorry dude. I'll get it updated in the Ruby lesson

**Review by CouchofTomato — CHANGES_REQUESTED** [review]

File: javascript/computer_science/time_complexity.md (hunk 1 above)
Comment (01zulfi) [reply javascript/computer_science/time_complexity.md:123]: This is intentional. I don't see this snippet as actual "code" but just a representation of an array. So it doesn't matter if we follow preferred JS practices. I don't feel strongly about this, I can change it if you want.

Comment (CouchofTomato) [reply javascript/computer_science/time_complexity.md:123]: No need. All good.

**Review by CouchofTomato — APPROVED** [review]

**Pull request merged**